---
title: "LC multilevel modeling"
author: "Molly"
date: "March 29, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
### For this analysis - we need to merge several data sets. The SNP data, the DHS individual data and the DHS cluster level ddta. We need to do this for the 2007 and the 2013 data, so we start with five different data sets and end with one final data set 

## loading in required libraries
library(lme4)
library(tidyverse)
library(dplyr)
library(haven)
library(readxl)
#install.packages("nnet")
library(nnet)

# start with reading in the combo data - a data set that has the cluster level covariates for both 2007 and 2013

combo <- read.csv("~/Documents/UNC/Meshnick Lab /DRC DHS data/Loved Children data/COMBINED.csv")

#########################
######## 2013 data ######
########################


## reading in the 2013 kids data set and the kids DHS covariates 
mipfull2 <- read_tsv("~/Desktop/mutation_summary.tsv")
DHSChildrenU5 <- read.csv("~/Documents/UNC/Meshnick Lab /DRC DHS data/Kids databases/Copy of DataFullKids.csv")

## subsetting only the DHPS SNPs 

dhps_sub <- subset(mipfull2, mipfull2$POS ==549685 | mipfull2$POS == 550117 | mipfull2$POS == 549993 | mipfull2$POS == 403625)
dhps_gt <- dhps_sub[c(1,5,52,54)]

## 
dhps_nomiss <- subset(dhps_gt, !is.na(dhps_gt$`Binary Genotype Call`))
A437G <- subset(dhps_nomiss, dhps_nomiss$POS == 549685)
A437G$A437G[A437G$`Binary Genotype Call` == "MUT"] <-1
A437G$A437G[A437G$`Binary Genotype Call` == "WT"] <-0

K540E <- subset(dhps_nomiss, dhps_nomiss$POS == 549993)
K540E$K540E[K540E$`Binary Genotype Call` == "MUT"] <-1
K540E$K540E[K540E$`Binary Genotype Call` == "WT"] <-0

A581G <- subset(dhps_nomiss, dhps_nomiss$POS == 550117)
A581G$A581G[A581G$`Binary Genotype Call` == "MUT"] <-1
A581G$A581G[A581G$`Binary Genotype Call` == "WT"] <-0

CRT <- subset(dhps_nomiss, dhps_nomiss$POS == 403625)
CRT$CRT[CRT$`Binary Genotype Call` == "MUT"] <-1
CRT$CRT[CRT$`Binary Genotype Call` == "WT"] <-0

dhps_merged <- merge(A437G, K540E, by = "Sample ID")
dhps_merged <- merge(dhps_merged, A581G, by = "Sample ID")
## making a mixed infection variable 
dhps_merged$mixed <- apply(dhps_merged, 1, function(x){as.numeric(any(x == "MIX", na.rm = T))}) #shit I love this function 

## subsetting just the SNP calls 
dhps_snps <- dhps_merged[c("Sample ID", "A437G", "K540E", "A581G", "mixed")]
dhps_snps <- merge(dhps_snps, CRT, by = "Sample ID")
dhps_snps <- dhps_merged[c("Sample ID", "A437G", "K540E", "A581G", "mixed", "CRT")]


dhps_snps$samplename <- sub("-SM-1", "", dhps_snps$`Sample ID`, fixed = T)
dhps_snps$Barcode <- str_sub(dhps_snps$samplename, -5)

### need to drop the controls 
dhps_snps <- merge(dhps_snps, DHSChildrenU5, by.x="Barcode", by.y = "sh312")


```

```{r}
############################
###### 2007 data ##########
###########################

## reading in adult 2007 data 
MoST_Congo <- read_dta("~/Documents/UNC/Meshnick Lab /DRC DHS data/2007 data/MoST-Congo.dta")
dhps_only <- readxl::read_xlsx("~/Documents/UNC/Meshnick Lab /DRC DHS data/2007 data/adult_dhps_only.xlsx")

## merging both 2007 data sets 
adult_merged <- merge(dhps_only, MoST_Congo, by.x = "barcode")

# subsetting only the SNP and cluster columns 
## keeping individual covariates we think are important
# wealth index 
dhs_sub <- adult_merged[c(1:6, 795, 807)]


## changing SNP coding to be NA,0,1 
## also changing column name so it matches 2013
dhs_sub$A437G[dhs_sub$d437 == 0] <- NA
dhs_sub$A437G[dhs_sub$d437 == 1] <- 0
dhs_sub$A437G[dhs_sub$d437 == 2 | dhs_sub$d437 == 3] <- 1

dhs_sub$K540E[dhs_sub$d540 == 0] <- NA
dhs_sub$K540E[dhs_sub$d540 == 1] <- 0
dhs_sub$K540E[dhs_sub$d540 == 2 | dhs_sub$d540 == 3] <- 1


dhs_sub$A581G[dhs_sub$d581 == 0] <- NA
dhs_sub$A581G[dhs_sub$d581 == 1] <- 0
dhs_sub$A581G[dhs_sub$d581 == 2 | dhs_sub$d581 == 3] <- 1

dhs_sub$crt_total[dhs_sub$pfcrt == 0] <- NA
dhs_sub$crt_total[dhs_sub$pfcrt == 1] <- 0
dhs_sub$crt_total[dhs_sub$pfcrt == 2 | dhs_sub$pfcrt == 3] <- 1

#making a variable for mixed infections 
dhs_sub$mixed[dhs_sub$d437 == 3 | dhs_sub$d540 == 3 | dhs_sub$d581 == 3] <- 1
dhs_sub$mixed[dhs_sub$d437 != 3 & dhs_sub$d540 != 3 & dhs_sub$d581 != 3] <- 0

```

```{r}
## now selecting the variables for the MLM 
## keeping barcode, SNP data, mixed column, 
adult_ind<- dhs_sub[c(1,7:13)]

## hv270 is the wealth variable 
kids_ind<- dhps_snps[c("Barcode", "A437G", "K540E", "A581G", "mixed", "CRT", "hv001", "hv270")]

## changing some variable name so we can combine these 
colnames(kids_ind)[1]<- "barcode"
colnames(kids_ind)[7]<- "cluster"
colnames(kids_ind)[8]<- "wealthindex"

colnames(adult_ind)[7]<- "CRT"
```

```{r}
## now looking at some of the cluster level covariate 
## using dplyr to get cluster level malaria prevalence 

cluster13 <- DHSChildrenU5 %>% 
  dplyr::group_by(hv001) %>%
  dplyr::summarise(total=n(), positive = sum(result))

cluster13$prevalence <- cluster13$positive / cluster13$total

cluster07 <- MoST_Congo %>% 
  dplyr::group_by(cluster) %>%
  dplyr::summarise(total=n(), positive = sum(falciparum_final))

cluster07$prevalence <- cluster07$positive / cluster07$total

## mering the individual covariates with the cluster level covariates 
kids_cluster <- subset(combo, combo$year == 2013) 
kids_full <- merge(kids_ind, kids_cluster, by ="cluster")
kids_full <- merge(kids_full, cluster13, by.x= "cluster", by.y="hv001")

adult_cluster <- subset(combo, combo$year == 2007)
adult_full <- merge(adult_ind, adult_cluster, by="cluster")
adult_full <- merge(adult_full, cluster07, by.x="cluster")
  # final adult data 
```

```{r}
# staggaring the cluster variable so that 2007 and 2013 dont overlap 
kids_full$cluster = kids_full$cluster + 0.5

# binding it all together for our final combined data set 
ind_full <- rbind(kids_full, adult_full)
```

```{r}
# okay! Now we have a combined data set -- need to make a joint dhps mutation outcome variable 

ind_full$dhps_mut[ind_full$A437G == 1 | ind_full$A581G == 1 | ind_full$K540E == 1] <-1 
ind_full$dhps_mut[ind_full$A437G == 0 & ind_full$A581G == 0 & ind_full$K540E == 0] <-0 

# also going to make a binary year variable 

ind_full$yearbin[ind_full$year == 2007] <- 0
ind_full$yearbin[ind_full$year == 2013] <- 1
```

```{r}
### looking at some basic summary stats 
dhps_tab <- table(ind_full$dhps_mut)
dhps_tab

dhps_year <- table(ind_full$dhps_mut, ind_full$year)
dhps_year

crt_tab <- table(ind_full$CRT, ind_full$year)
crt_tab

crt_dhps <- table(ind_full$CRT, ind_full$dhps_mut)
crt_dhps

tapply(ind_full$FansPercent, ind_full$yearbin, summary) # Fansidar use increased 
tapply(ind_full$prevalence, ind_full$yearbin, summary) # mean prevalence also increased 
mixed_tab <- table(ind_full$mixed, ind_full$year)
mixed_tab


```

```{r}
## woahhhhh okay finally on to the modeling part 
## using the lmer package and grouping by year and cluster 
## joint outcome for any dhps mutaion 

m_full <- lmer(dhps_mut ~ prevalence 
             + FansPercent 
             + wealth_mean
             + urban1rural0
             + pctedu0
             + mixed
             + factor(wealthindex)
             + (1|year/cluster), data=ind_full)
summary(m_full)

## from this model - wealth has the highest p-value so we will drop 

m_red1 <- lmer(dhps_mut ~ prevalence 
             + FansPercent 
#             + wealth_mean
             + urban1rural0
             + pctedu0
             + (1|year/cluster), data=nomiss2)
summary(m_red1)

m_red2 <- lmer(dhps_mut ~  FansPercent 
             + urban1rural0
             + pctedu0
             + (1|year/cluster), data=nomiss2)
summary(m_red2)

m_red3 <- lmer(dhps_mut ~  FansPercent 
             + urban1rural0
             + (1|year/cluster), data=nomiss2)
summary(m_red3)
```


```{r}
## now looking only at the K540E mutation 

k_full <- lmer(K540E ~ prevalence 
             + FansPercent 
             + wealth_mean
             + urban1rural0
             + pctedu0
             + mixed 
             + factor(wealthindex)
             + (1|year/cluster), data=ind_full)
summary(k_full)
AIC(k_full)


k_red1 <- lmer(K540E ~ prevalence 
             + FansPercent 
             + wealth_mean
             + urban1rural0
             + mixed
             + factor(wealthindex)
             + (1|year/cluster), data=ind_full)
summary(k_red1)
AIC(k_red1)

k_red2 <- lmer(K540E ~ prevalence 
             + FansPercent 
             + wealth_mean
             + mixed
             + factor(wealthindex)
             + (1|year/cluster), data=ind_full)
summary(k_red2)
AIC(k_red2)

```

```{r}
## trying some univariate models 
## any mutation outcome 
uni1 <- glmer(dhps_mut  ~ prevalence + (1|year/cluster), data = ind_full, family=binomial(link = logit))
summary(uni1)

uni2 <- glmer(dhps_mut  ~ FansPercent+ (1|year/cluster), data = ind_full, family=binomial(link = logit))
summary(uni2)

uni3 <- glmer(dhps_mut  ~ mixed + (1|year/cluster), data = ind_full, family=binomial(link = logit))
summary(uni3)

uni4 <- glmer(dhps_mut  ~ pctedu0+ (1|year/cluster), data = ind_full, family=binomial(link = logit))
summary(uni4)

uni5 <- glmer(dhps_mut  ~ urban1rural0 + (1|year/cluster), data = ind_full, family=binomial(link = logit))
summary(uni5)

uni6 <- glmer(dhps_mut  ~ wealthindex + (1|year/cluster), data = ind_full, family=binomial(link = logit))
summary(uni6)

uni7 <- glmer(dhps_mut  ~ wealth_mean + (1|year/cluster), data = ind_full, family=binomial(link = logit))
summary(uni7)
```

```{r}
### 540E outcome 

uni_k1 <- glmer(K540E  ~ prevalence + (1|year/cluster), data = ind_full, family=binomial(link = logit))
summary(uni_k1)

uni_k2 <- glmer(K540E  ~ FansPercent + (1|year/cluster), data = ind_full, family=binomial(link = logit))
summary(uni_k2)

uni_k3 <- glmer(K540E  ~ wealthindex + (1|year/cluster), data = ind_full, family=binomial(link = logit))
summary(uni_k3)

uni_k4 <- glmer(K540E  ~ wealth_mean + (1|year/cluster), data = ind_full, family=binomial(link = logit))
summary(uni_k4)

uni_k5 <- glmer(K540E  ~ mixed + (1|year/cluster), data = ind_full, family=binomial(link = logit))
summary(uni_k5)

uni_k6 <- glmer(K540E  ~ pctedu0 + (1|year/cluster), data = ind_full, family=binomial(link = logit))
summary(uni_k6)

uni_k7 <- glmer(K540E  ~ urban1rural0 + (1|year/cluster), data = ind_full, family=binomial(link = logit))
summary(uni_k7)
```

```{r}
## making a factor variable 
ind_full$haplotype[ind_full$A437G == 0 & ind_full$K540E == 0 & ind_full$A581G == 0] <- "aaa"
ind_full$haplotype[ind_full$A437G == 1 & ind_full$K540E == 0 & ind_full$A581G == 0] <- "GKA"
ind_full$haplotype[ind_full$A437G == 0 & ind_full$K540E == 1 & ind_full$A581G == 0] <- "AEA"
ind_full$haplotype[ind_full$A437G == 0 & ind_full$K540E == 1 & ind_full$A581G == 1] <- "AEG"
ind_full$haplotype[ind_full$A437G == 1 & ind_full$K540E == 1 & ind_full$A581G == 1] <- "GEG"
ind_full$haplotype[ind_full$A437G == 1 & ind_full$K540E == 1 & ind_full$A581G == 0] <- "GEA"


poly_1 <- multinom(haplotype  ~ prevalence + (1|year/cluster), data = ind_full, family=binomial(link = logit))
summary(poly_1)
?multinom
```



